packages/cli/src/lib/parallel.ts

```diff
@@ -14,34 +14,79 @@
  * limitations under the License.
  */
 
-export const PARALLEL_ENV_VAR = 'BACKSTAGE_CLI_BUILD_PARALLEL';
+export const DEFAULT_PARALLELISM = 4;
 
-export type ParallelOption = boolean | number | undefined;
+export const PARALLEL_ENV_VAR = 'BACKSTAGE_CLI_BUILD_PARALLEL';
 
-export function isParallelDefault(parallel: ParallelOption) {
-  return parallel === undefined || parallel === true;
-}
+export type ParallelismOption = boolean | string | number | null | undefined;
 
-export function parseParallel(
-  parallel: boolean | string | number | undefined,
-): ParallelOption {
+export function parseParallelismOption(parallel: ParallelismOption): number {
   if (parallel === undefined || parallel === null) {
-    return true;
+    return DEFAULT_PARALLELISM;
   } else if (typeof parallel === 'boolean') {
-    return parallel;
+    return parallel ? DEFAULT_PARALLELISM : 1;
   } else if (typeof parallel === 'number' && Number.isInteger(parallel)) {
+    if (parallel < 1) {
+      return 1;
+    }
     return parallel;
   } else if (typeof parallel === 'string') {
     if (parallel === 'true') {
-      return true;
+      return parseParallelismOption(true);
     } else if (parallel === 'false') {
-      return false;
-    } else if (Number.isInteger(parseFloat(parallel.toString()))) {
-      return Number(parallel);
+      return parseParallelismOption(false);
+    }
+    const parsed = Number(parallel);
+    if (Number.isInteger(parsed)) {
+      return parseParallelismOption(parsed);
     }
   }
 
   throw Error(
     `Parallel option value '${parallel}' is not a boolean or integer`,
   );
 }
+
+export function getEnvironmentParallelism() {
+  return parseParallelismOption(process.env[PARALLEL_ENV_VAR]);
+}
+
+type ParallelWorkerOptions<TItem> = {
+  /**
+   * Decides the number of parallel workers by multiplying
+   * this with the configured parallelism, which defaults to 4.
+   *
+   * Defaults to 1.
+   */
+  parallelismFactor?: number;
+  parallelismSetting?: ParallelismOption;
+  items: Iterable<TItem>;
+  worker: (item: TItem) => Promise<void>;
+};
+
+export async function runParallelWorkers<TItem>(
+  options: ParallelWorkerOptions<TItem>,
+) {
+  const { parallelismFactor = 1, parallelismSetting, items, worker } = options;
+  const parallelism = parallelismSetting
+    ? parseParallelismOption(parallelismSetting)
+    : getEnvironmentParallelism();
+
+  const iterator = items[Symbol.iterator]();
+
+  async function pop() {
+    const el = iterator.next();
+    if (el.done) {
+      return;
+    }
+
+    await worker(el.value);
+    await pop();
```

Comment:
does v8 do any tail call optimization? ;)

feels like this could just have been a loop instead that breaks when the iterator is done, instead of calling oneself again

Reply:
Yeah could prolly be refactored into that. Didn't look much at the implementation itself tbh, just moved it over.